src/riscv/linux/init.c

```diff
@@ -227,75 +257,95 @@ void cpuinfo_riscv_linux_init(void) {
 	struct cpuinfo_cluster* clusters = NULL;
 	struct cpuinfo_core* cores = NULL;
 	struct cpuinfo_uarch_info* uarchs = NULL;
+	struct cpuinfo_cache* l1i = NULL;
+	struct cpuinfo_cache* l1d = NULL;
+	struct cpuinfo_cache* l2 = NULL;
 	const struct cpuinfo_processor** linux_cpu_to_processor_map = NULL;
 	const struct cpuinfo_core** linux_cpu_to_core_map = NULL;
 	uint32_t* linux_cpu_to_uarch_index_map = NULL;
 
-	/**
-	 * The interesting set of processors are the number of 'present'
-	 * processors on the system. There may be more 'possible' processors, but
-	 * processor information cannot be gathered on non-present processors.
-	 *
-	 * Note: For SoCs, it is largely the case that all processors are known
-	 * at boot and no processors are hotplugged at runtime, so the
-	 * 'present' and 'possible' list is often the same.
-	 *
-	 * Note: This computes the maximum processor ID of the 'present'
-	 * processors. It is not a count of the number of processors on the
-	 * system.
-	 */
-	const uint32_t max_processor_id = 1 +
-		cpuinfo_linux_get_max_present_processor(
-				cpuinfo_linux_get_max_processors_count());
-	if (max_processor_id == 0) {
-		cpuinfo_log_error("failed to discover any processors");
+	const uint32_t max_processors_count = cpuinfo_linux_get_max_processors_count();
+	cpuinfo_log_debug("system maximum processors count: %"PRIu32, max_processors_count);
+
+	const uint32_t max_possible_processors_count = 1 +
+                                                   cpuinfo_linux_get_max_possible_processor(max_processors_count);
+	cpuinfo_log_debug("maximum possible processors count: %"PRIu32, max_possible_processors_count);
+	const uint32_t max_present_processors_count = 1 +
+                                                  cpuinfo_linux_get_max_present_processor(max_processors_count);
+	cpuinfo_log_debug("maximum present processors count: %"PRIu32, max_present_processors_count);
+
+	uint32_t valid_processor_mask = 0;
+	uint32_t riscv_linux_processors_count = max_processors_count;
+	if (max_present_processors_count != 0) {
+		riscv_linux_processors_count = min(riscv_linux_processors_count, max_present_processors_count);
+		valid_processor_mask |= CPUINFO_LINUX_FLAG_PRESENT;
+	}
+	if (max_possible_processors_count != 0) {
+		riscv_linux_processors_count = min(riscv_linux_processors_count, max_possible_processors_count);
+		valid_processor_mask |= CPUINFO_LINUX_FLAG_POSSIBLE;
+	}
+	if ((max_present_processors_count | max_possible_processors_count) == 0) {
+		cpuinfo_log_error("failed to parse both lists of possible and present processors");
```

Comment:
I'm not quite following why the ARM code is written in the same way - in what case do we need to mix the present and possible processors, given that we can only ever infer data about the present processors?